include/boost/beast/core/impl/buffers_adaptor.hpp

```diff
@@ -34,332 +35,162 @@ namespace beast {
 
 template<class MutableBufferSequence>
 template<bool isMutable>
-class buffers_adaptor<MutableBufferSequence>::
-    readable_bytes
+class buffers_adaptor<MutableBufferSequence>::subrange
 {
-    buffers_adaptor const* b_;
-
 public:
-    using value_type = typename
-        std::conditional<isMutable,
-            net::mutable_buffer,
-            net::const_buffer>::type;
+    using value_type = typename std::conditional<
+        isMutable,
+        net::mutable_buffer,
+        net::const_buffer>::type;
 
-    class const_iterator;
+    struct iterator;
 
-    readable_bytes() = delete;
+    // construct from two iterators plus optionally subrange definition
+    subrange(
+        iter_type first,        // iterator to first buffer in storage
+        iter_type last,         // iterator to last buffer in storage
+        std::size_t pos = 0,    // the offset in bytes from the beginning of the storage
+        std::size_t n =         // the total length of the subrange
+        (std::numeric_limits<std::size_t>::max)());
 
 #if BOOST_WORKAROUND(BOOST_MSVC, < 1910)
-    readable_bytes(
-        readable_bytes const& other)
-        : b_(other.b_)
+    subrange(
+        subrange const& other)
+        : first_(other.first_)
+        , last_(other.last_)
+        , first_offset_(other.first_offset_)
+        , last_size_(other.last_size_)
     {
     }
 
     readable_bytes& operator=(
         readable_bytes const& other)
     {
-        b_ = other.b_;
+        first_ = other.first_;
+        last_ = other.last_;
+        first_offset_ = other.first_offset_;
+        last_size_ = other.last_size_;
         return *this;
     }
 #else
-    readable_bytes(
-        readable_bytes const&) = default;
-    readable_bytes& operator=(
-        readable_bytes const&) = default;
+    subrange(
+        subrange const&) = default;
+    subrange& operator=(
+        subrange const&) = default;
 #endif
 
-    template<bool isMutable_ = isMutable, class =
-        typename std::enable_if<! isMutable_>::type>
-    readable_bytes(
-        readable_bytes<true> const& other) noexcept
-        : b_(other.b_)
+    // allow conversion from mutable to const
+    template<bool thisIsMutable = isMutable,
+        typename std::enable_if<!thisIsMutable>::type * = nullptr>
+    subrange(subrange<true> const &other)
+        : first_(other.first())
+        , last_(other.last())
+        , first_offset_(other.first_offset())
+        , last_size_(other.last_size())
```

Comment:
These member functions `first()`, `last()`, `first_offset()`, and `last_size()` seem unnecessary, you could just access the data members. You will have to make the class template a friend first of course.